### PR TITLE
App-Lock follow-up

### PR DESCRIPTION
**Pull request by smatting**
Implements
https://wearezeta.atlassian.net/browse/SQSERVICES-80

Note before deploying: This PR requires changes to galley config!

### CODE REVIEW

**Comment by fisx** [comment]
can you please make config changes optional?  this makes migration easier in general, and will help us get this to staging more quickly.

File: services/galley/src/Galley/Data/TeamFeatures.hs [hunk 1]
```diff
@@ -87,10 +87,11 @@ getApplockFeatureStatus tid = do
   let q = query1 (select) (params Quorum (Identity tid))
   mTuple <- retry x1 q
   pure $
-    mTuple <&> \(statusValue, enforce, timeout) ->
-      TeamFeatureStatusWithConfig statusValue (Public.TeamFeatureAppLockConfig enforce timeout)
+    join $
+      mTuple <&> \(mbStatusValue, mbEnforce, mbTimeout) ->
```
Comment (fisx) [inline services/galley/src/Galley/Data/TeamFeatures.hs:91]:
```suggestion
    mTuple >>= \(mbStatusValue, mbEnforce, mbTimeout) ->
```

File: services/galley/src/Galley/API/Internal.hs [hunk 2]
```diff
@@ -336,17 +336,25 @@ mkFeatureGetAndPutRoute getter setter = do
...
-  get ("/i/teams/:tid/features/" <> toByteString' featureName) (continue getHandler) $
-    capture "tid"
-      .&. accept "application" "json"
+  let mkGetRoute name =
+        get ("/i/teams/:tid/features/" <> name) (continue getHandler) $
+          capture "tid"
+            .&. accept "application" "json"
+
+  mkGetRoute (toByteString' featureName)
+  mkGetRoute `mapM_` (Public.deprecatedFeatureName featureName)
```
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:345]:
```suggestion
  mkGetRoute `mapM_` Public.deprecatedFeatureName featureName
```

File: services/galley/src/Galley/API/Internal.hs [hunk 3]
```diff
@@ -336,17 +336,25 @@ mkFeatureGetAndPutRoute getter setter = do
...
-      .&. jsonRequest @(Public.TeamFeatureStatus a)
-      .&. accept "application" "json"
+  let mkPutRoute name =
+        put ("/i/teams/:tid/features/" <> name) (continue putHandler) $
+          capture "tid"
+            .&. jsonRequest @(Public.TeamFeatureStatus a)
+            .&. accept "application" "json"
+
+  mkPutRoute (toByteString' featureName)
+  mkPutRoute `mapM_` (Public.deprecatedFeatureName featureName)
```
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:360]:
```suggestion
  mkPutRoute `mapM_` Public.deprecatedFeatureName featureName
```

File: services/galley/src/Galley/API/Public.hs [hunk 4]
```diff
@@ -458,6 +458,16 @@ sitemap = do
   mkFeatureGetAndPutRoute @'Public.TeamFeatureDigitalSignatures Teams.getDigitalSignaturesInternal Teams.setDigitalSignaturesInternal
   mkFeatureGetAndPutRoute @'Public.TeamFeatureAppLock Teams.getAppLockInternal Teams.setAppLockInternal
 
+  get "/teams/:tid/features/" (continue Teams.getAllFeaturesH) $
```
Comment (fisx) [inline services/galley/src/Galley/API/Public.hs:461]:
```suggestion
  get "/teams/:tid/features" (continue Teams.getAllFeaturesH) $
```

(not sure it makes a difference, but this is how all other end-points in this module do it.)

File: services/galley/src/Galley/API/Teams.hs [hunk 5]
```diff
@@ -908,6 +911,36 @@ setFeatureStatus setter doauth tid status = do
...
+    <$> sequence
+      [ w @'Public.TeamFeatureSSO getSSOStatusInternal,
+        w @'Public.TeamFeatureLegalHold getLegalholdStatusInternal,
+        w @'Public.TeamFeatureSearchVisibility getTeamSearchVisibilityAvailableInternal,
+        w @'Public.TeamFeatureValidateSAMLEmails getValidateSAMLEmailsInternal,
+        w @'Public.TeamFeatureDigitalSignatures getDigitalSignaturesInternal,
+        w @'Public.TeamFeatureAppLock getAppLockInternal
+      ]
+  where
+    w ::
```
Comment (fisx) [inline services/galley/src/Galley/API/Teams.hs:930]: Please avoid one-letter identifiers.  (General rule, it's ok if you have no time to change it here now.)

File: services/galley/src/Galley/API/Teams.hs [hunk 6]
```diff
@@ -905,9 +908,39 @@ setFeatureStatus setter doauth tid status = do
...
+      forall (a :: Public.TeamFeatureName).
+      ( Public.KnownTeamFeatureName a,
+        Aeson.ToJSON (Public.TeamFeatureStatus a)
+      ) =>
+      ( TeamId ->
+        Galley (Public.TeamFeatureStatus a)
+      ) ->
+      Galley (Text, Aeson.Value)
```
Comment (fisx) [inline services/galley/src/Galley/API/Teams.hs:938]:
```suggestion
      (Public.KnownTeamFeatureName a, Aeson.ToJSON (Public.TeamFeatureStatus a)) =>
      (TeamId -> Galley (Public.TeamFeatureStatus a)) ->
      Galley (Aeson.Pair)
```

File: services/galley/src/Galley/API/Teams.hs [hunk 7]
```diff
@@ -905,9 +908,39 @@ setFeatureStatus setter doauth tid status = do
...
+    w getter = do
+      status <- getFeatureStatus @a getter (DoAuth uid) tid
+      let feature = Public.knownTeamFeatureName @a
+      pure $ (cs . toByteString' $ feature, Aeson.toJSON status)
```
Comment (fisx) [inline services/galley/src/Galley/API/Teams.hs:942]:
```suggestion
      pure $ (feature Aeson..= status)
```

(This requires a `KeyValue` instance for `TeamFeatureName`.)

File: deploy/services-demo/conf/galley.demo.yaml [hunk 8]
```diff
@@ -33,6 +33,12 @@ settings:
     sso: disabled-by-default
     legalhold: disabled-by-default
     teamSearchVisibility: disabled-by-default
+    appLock:
+      defaults:
+        status: enabled
+        config:
+          enforceAppLock: false
+          inactivityTimeoutSecs: 60
```
Comment (fisx) [inline deploy/services-demo/conf/galley.demo.yaml:41]:
```suggestion
      status: enabled
      config:
        enforceAppLock: false
        inactivityTimeoutSecs: 60
```

(same with galley.demo-docker.yaml)

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 9]
```diff
@@ -196,10 +198,22 @@ newtype TeamCreationTime = TeamCreationTime
 data FeatureFlags = FeatureFlags
   { _flagSSO :: !FeatureSSO,
     _flagLegalHold :: !FeatureLegalHold,
-    _flagTeamSearchVisibility :: !FeatureTeamSearchVisibility
+    _flagTeamSearchVisibility :: !FeatureTeamSearchVisibility,
+    _flagAppLockDefaults :: !(Defaults (TeamFeatureStatus 'TeamFeatureAppLock))
   }
   deriving (Eq, Show, Generic)
 
+newtype Defaults a = Defaults {_unDefaults :: a}
```
Comment (fisx) [inline libs/galley-types/src/Galley/Types/Teams.hs:206]: I think this type can go away entirely.

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 10]
```diff
@@ -196,10 +198,22 @@ newtype TeamCreationTime = TeamCreationTime
 data FeatureFlags = FeatureFlags
   { _flagSSO :: !FeatureSSO,
     _flagLegalHold :: !FeatureLegalHold,
-    _flagTeamSearchVisibility :: !FeatureTeamSearchVisibility
+    _flagTeamSearchVisibility :: !FeatureTeamSearchVisibility,
+    _flagAppLockDefaults :: !(Defaults (TeamFeatureStatus 'TeamFeatureAppLock))
```
Comment (fisx) [inline libs/galley-types/src/Galley/Types/Teams.hs:202]:
```suggestion
    _flagAppLock :: !(Defaults (TeamFeatureStatus 'TeamFeatureAppLock))
```

(for consistency with the other flags)

**Review by fisx — COMMENTED** [review]
actually, not sure about the config thing i said any more.  your approach is very elegant.  i'll leave in my comments for now, let's fight over this!  :)

**Comment by fisx** [comment]
ah, and please run all the test.  or at least compile everything.  :)

```
galley-types                > /home/mf/src/wire-server/libs/galley-types/test/unit/Test/Galley/Types.hs:63:5: error:
galley-types                >     • Couldn't match type ‘Defaults
galley-types                >                              (Wire.API.Team.Feature.TeamFeatureStatus
galley-types                >                                 'Wire.API.Team.Feature.TeamFeatureAppLock)
galley-types                >                            -> FeatureFlags’
galley-types                >                      with ‘FeatureFlags’
galley-types                >       Expected type: FeatureSSO  
galley-types                >                      -> FeatureLegalHold -> FeatureTeamSearchVisibility -> FeatureFlags
galley-types                >         Actual type: FeatureSSO  
galley-types                >                      -> FeatureLegalHold
galley-types                >                      -> FeatureTeamSearchVisibility
galley-types                >                      -> Defaults 
galley-types                >                           (Wire.API.Team.Feature.TeamFeatureStatus
galley-types                >                              'Wire.API.Team.Feature.TeamFeatureAppLock)
galley-types                >                      -> FeatureFlags
galley-types                >     • Probable cause: ‘FeatureFlags’ is applied to too few arguments
galley-types                >       In the first argument of ‘(<$>)’, namely ‘FeatureFlags’
galley-types                >       In the first argument of ‘(<*>)’, namely
galley-types                >         ‘FeatureFlags <$> QC.elements [minBound .. ]’
galley-types                >       In the first argument of ‘(<*>)’, namely
galley-types                >         ‘FeatureFlags <$> QC.elements [minBound .. ]
galley-types                >            <*> QC.elements [minBound .. ]’
galley-types                >    |                             
galley-types                > 63 |     FeatureFlags            
galley-types                >    |     ^^^^^^^^^^^^            
```

File: libs/galley-types/src/Galley/Types/Teams.hs (hunk 9 above)
Comment (smatting) [reply libs/galley-types/src/Galley/Types/Teams.hs]: discussed over call

File: libs/galley-types/src/Galley/Types/Teams.hs (hunk 10 above)
Comment (smatting) [reply libs/galley-types/src/Galley/Types/Teams.hs]: resolved in call

File: services/galley/src/Galley/Data/TeamFeatures.hs (hunk 1 above)
Comment (smatting) [reply services/galley/src/Galley/Data/TeamFeatures.hs:91]: done

File: services/galley/src/Galley/API/Public.hs (hunk 4 above)
Comment (smatting) [reply services/galley/src/Galley/API/Public.hs:461]: consistency is good

File: services/galley/src/Galley/API/Teams.hs (hunk 7 above)
Comment (smatting) [reply services/galley/src/Galley/API/Teams.hs:942]: applied your suggestion partly. Looks like first argument to `.=` is not overloaded, but must be `Text`

File: deploy/services-demo/conf/galley.demo.yaml (hunk 8 above)
Comment (smatting) [reply deploy/services-demo/conf/galley.demo.yaml]: resolved in call. almost decided to keep it that way

**Comment by smatting** [comment]
@fisx approved

File: services/galley/src/Galley/API/Public.hs (hunk 4 above)
Comment (fisx) [reply services/galley/src/Galley/API/Public.hs:461]: doesn't look like this made it to develop?  @smatting and it looks like the trailing `/` is mandatory if you have it there.  m-|

**Pull request merged**